### ISSUE TITLE
add_host seems to not be working?

### ISSUE THREAD

**Issue by mark-casey** [ansible/ansible#5791]
This is a simplified version...(original: https://groups.google.com/forum/#!topic/ansible-project/c3pty9IrqRc). Am I doing something wrong?

ansible 1.4.3

``` yaml
- name: Group adding test
  hosts: localhost

  tasks:
    - name: Add a group
      add_host: hostname=localhost groups=testgroup

    - name: See if it is there
      debug: var=group_names
```

```


jenkins@ip-172-31-25-124:~/jobs/JobName/workspace$ ansible-playbook -vvvv -i 'localhost,' test.yml --connection=local

PLAY [Group adding test] ******************************************************

GATHERING FACTS ***************************************************************
<localhost> EXEC ['/bin/sh', '-c', 'mkdir -p $HOME/.ansible/tmp/ansible-1390856983.1-251208038321775 && chmod a+rx $HOME/.ansible/tmp/ansible-1390856983.1-251208038321775 && echo $HOME/.ansible/tmp/ansible-1390856983.1-251208038321775']
<localhost> REMOTE_MODULE setup
<localhost> PUT /tmp/tmpMgWuqZ TO /var/lib/jenkins/.ansible/tmp/ansible-1390856983.1-251208038321775/setup
<localhost> EXEC ['/bin/sh', '-c', '/usr/bin/python /var/lib/jenkins/.ansible/tmp/ansible-1390856983.1-251208038321775/setup; rm -rf /var/lib/jenkins/.ansible/tmp/ansible-1390856983.1-251208038321775/ >/dev/null 2>&1']
ok: [localhost]

TASK: [Add a group] ***********************************************************
creating host via 'add_host': hostname=localhost
added host to group via add_host module: testgroup
ok: [localhost] => {"new_groups": ["testgroup"], "new_host": "localhost"}

TASK: [See if it is ther] *****************************************************
ok: [localhost] => {
    "group_names": []
}

PLAY RECAP ********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0
```

**Comment by jctanner** [comment]
@mark-casey in this case localhost is being handled in a special way. If you add a host called "foo" and start a new play, you'll see the expected behavior.

``` yaml
- hosts: localhost
  connection: local
  gather_facts: False
  tasks:
    - name: Add a group
      add_host: hostname=foo groups=testgroup

    - name: See if it is there
      debug: var=group_names

    - debug: var=hostvars

- hosts: testgroup
  connection: local
  gather_facts: false
  tasks:
    - debug: var=group_names
    - debug: msg="{{ inventory_hostname }}"
    - debug: var=hostvars[inventory_hostname]
```

```
PLAY [localhost] **************************************************************

TASK: [Add a group] ***********************************************************
ok: [localhost]

TASK: [See if it is there] ****************************************************
ok: [localhost] => {
    "group_names": [
        "ungrouped"
    ]
}

TASK: [debug var=hostvars] ****************************************************
ok: [localhost] => {
    "hostvars": {
        "localhost": {}
    }
}

PLAY [testgroup] **************************************************************

TASK: [debug var=group_names] *************************************************
ok: [foo] => {
    "group_names": [
        "testgroup"
    ]
}

TASK: [debug msg="{{inventory_hostname}}"] ************************************
ok: [foo] => {
    "msg": "foo"
}

TASK: [debug var=hostvars[inventory_hostname]] ********************************
ok: [foo] => {
    "hostvars[inventory_hostname]": {
        "group_names": [
            "testgroup"
        ],
        "inventory_hostname": "foo",
        "inventory_hostname_short": "foo"
    }
}

PLAY RECAP ********************************************************************
foo                        : ok=3    changed=0    unreachable=0    failed=0
localhost                  : ok=3    changed=0    unreachable=0    failed=0
```

**Comment by mpdehaan** [comment]
I'm not sure why it should be held to a different standard here.

It may not be a real world problem in most cases though.

**Comment by mark-casey** [comment]
Yes, my original issue on-list was using add_host with ec2 instances (where I'm having similar trouble getting the new groups to show up); I just used localhost in an attempt to be more succinct.

However, FWIW I have run into this localhost thing before. (at least, based on my understanding of it I have... I know its still possible I'm missing something **obvious** across the board here)

We use vagrant with virtualbox for our development environment and I wanted to deploy it with the same roles use in prod on ec2. So I wrote a quick dummy_iaas role that just uses add_host to add localhost to the same provider-agnostic groups that I'm having trouble getting to show up when working with ec2. Then all of the selected roles like web and db just deploy on one local virtualbox.

So as it happens, even though localhost doesn't show up in group_names after the add_host, you can still apply deployment roles to it with plays that target the groups it supposedly isn't a part of. If you change the word 'foo' back to localhost in James' example above, localhost will trigger for the second play as if it is a part of testgroup...and then it will debug print that it isn't in the group at all...

```
- hosts: localhost
  connection: local
  gather_facts: False
  tasks:
    - name: Add a group
      add_host: hostname=localhost groups=testgroup

    - name: See if it is there
      debug: var=group_names

    - debug: var=hostvars

- hosts: testgroup
  connection: local
  gather_facts: false
  tasks:
    - debug: var=group_names
    - debug: msg="{{ inventory_hostname }}"
    - debug: var=hostvars[inventory_hostname]
```

```
PLAY [localhost] **************************************************************

TASK: [Add a group] ***********************************************************
creating host via 'add_host': hostname=localhost
added host to group via add_host module: testgroup
ok: [localhost] => {"new_groups": ["testgroup"], "new_host": "localhost"}

TASK: [See if it is there] ****************************************************
ok: [localhost] => {
    "group_names": []
}

TASK: [debug var=hostvars] ****************************************************
ok: [localhost] => {
    "hostvars": {
        "localhost": {}
    }
}

PLAY [testgroup] **************************************************************

TASK: [debug var=group_names] *************************************************
ok: [localhost] => {
    "group_names": []
}

TASK: [debug msg="{{inventory_hostname}}"] ************************************
ok: [localhost] => {
    "msg": "localhost"
}

TASK: [debug var=hostvars[inventory_hostname]] ********************************
ok: [localhost] => {
    "hostvars[inventory_hostname]": {
        "group_names": [],
        "inventory_hostname": "localhost",
        "inventory_hostname_short": "localhost"
    }
}

PLAY RECAP ********************************************************************
localhost                  : ok=6    changed=0    unreachable=0    failed=0
```

So it doesn't hurt me that localhost's new groups don't show in group_names but it would kill me dead if plays suddenly stopped applying to the groups it thinks it isn't a member of.

**Comment by mpdehaan** [comment]
I'm having difficulty understanding much of the above, but add_host we know to be operational.

In the above the foo group is added to testgroup and the output configures testgroup, which looks correct to me.

**Comment by mark-casey** [comment]
Yea, in re-reading some of that it is hard to follow... basically add_host wasn't doing what expected either in ec2 or for localhost. The ec2 issue was my fault but I came to the conclusion that the localhost problem was a real thing.

I've since worked around the problem I was seeing here and I can simply accept that it is just a thing that only applies to localhost. As I can still reproduce it though, I do want to leave it with one more stab at a succinct description, if just for someone in the future.

The core issue was that when you run James' example, this text is in the play's output:

```
TASK: [debug var=group_names] *************************************************
ok: [foo] => {
    "group_names": [
        "testgroup"
    ]
}
```

So host foo is a member of testgroup, and this is correct.

However if you change his add_host line to:

```
add_host: hostname=localhost groups=testgroup
```

and re-run the play, then this is in the play's output instead:

```
TASK: [debug var=group_names] *************************************************
ok: [localhost] => {
    "group_names": [
        "ungrouped"
    ]
}
```

So during each run a host was added to testgroup and during each run that host executed a play targeted at testgroup... but only one of them actually thinks it is a member of the group when you debug print.
